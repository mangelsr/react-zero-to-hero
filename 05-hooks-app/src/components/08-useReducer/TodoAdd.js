import React from 'react'
import { useForm } from '../../hooks/useForm';


export const TodoAdd = ({handleAddTodo}) => {

    const [ { desc }, handleInputChange, reset ] = useForm({ desc: '' });

    const handleSubmit = (e) => {
        e.preventDefault();

        if (desc.trim().length <= 1) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc,
            done: false
        }

        handleAddTodo(newTodo);

        reset();
    };

    return (
        <>
            <h4>Agregar TODO</h4>
            <hr/>
            <form onSubmit={handleSubmit}>
                <input 
                    type="text"
                    name="desc"
                    placeholder="Aprender..."
                    autoComplete="off"
                    className="form-control"
                    value={desc}
                    onChange={handleInputChange}
                />
                <button 
                    type="submit"
                    className="btn btn-outline-primary btn-block mt-2 ">
                    Agregar
                </button>
            </form>
        </>
    )
}
