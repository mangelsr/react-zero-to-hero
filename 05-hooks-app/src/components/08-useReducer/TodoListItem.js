import React from 'react'


export const TodoListItem = ({todo, i, handleDelete, handleToggle}) => {
    return (
        <li className="list-group-item">
            <p
                onClick={() => handleToggle(todo.id)}
                className={`${todo.done && 'done'}`}>
                {i+1}. {todo.desc}
            </p>
            <button
                onClick={() => handleDelete(todo.id)}
                className="btn btn-danger">
                Delete
            </button>
        </li>
    )
}
