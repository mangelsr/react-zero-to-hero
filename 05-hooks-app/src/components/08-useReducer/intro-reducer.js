console.log('Hola mundo');

const initialState = [{
    id: 1,
    todo: 'Comprar un PS4',
    done: false
}];

const todoReducer = (state = initialState, action) => {
    switch (action?.type) {
        case 'ADD':
            return [...state, action.payload];
        default:
            return state;
    }
}

let todos = todoReducer();

const newTodo = {
    id: 2,
    todo: 'Comprar KH3',
    done: false
};

const addTodoAction = {
    type: 'ADD', // standar
    payload: newTodo // standar para argumentos
};

todos = todoReducer(todos, addTodoAction);




console.log(todos);