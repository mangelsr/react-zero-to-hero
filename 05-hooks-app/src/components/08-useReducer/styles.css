p {
    margin: auto 0;
    cursor: pointer;
}

.list-group-item {
    display: flex;
    align-items: center;
    justify-content: space-between;
}

.done {
    text-decoration: line-through;
}