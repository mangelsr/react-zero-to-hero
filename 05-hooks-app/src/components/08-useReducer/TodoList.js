import React from 'react'

import { TodoListItem } from './TodoListItem'


export const TodoList = ({todos, handleDelete, handleToggle}) => {
    return (
        <ul className="list-group list-group-flush">
            {todos.map((todo, i) => 
                <TodoListItem
                    key={todo.id}
                    i={i}
                    todo={todo}
                    handleToggle={handleToggle}
                    handleDelete={handleDelete}
                />
            )}
        </ul>
    )
}
