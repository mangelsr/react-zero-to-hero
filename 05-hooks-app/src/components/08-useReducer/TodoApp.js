import React, { useEffect, useReducer } from 'react'

import { todoReducer } from './todoReducer';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';
import './styles.css';


const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];
};


export const TodoApp = () => {

    const [ todos, dispatch ] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);

    const handleDelete = (todoId) => {
        const action = {
            type: 'DELETE',
            payload: todoId
        };
        dispatch(action);
    };

    const handleToggle = (todoId) => {
        const action = {
            type: 'TOGGLE',
            payload: todoId
        };
        dispatch(action);
    };

    const handleAddTodo = (newTodo) => {
        dispatch({
            type: 'ADD',
            payload: newTodo
        });
    };

    return (
        <div>
            <h1>Todo App ({todos.length})</h1>
            <hr/>
            <div className="row">
                <div className="col-7">
                    <TodoList
                        todos={todos}
                        handleToggle={handleToggle}
                        handleDelete={handleDelete}
                    />
                </div>
                <div className="col-5">
                    <TodoAdd handleAddTodo={handleAddTodo} />
                </div>
            </div>
        </div>
    )
}