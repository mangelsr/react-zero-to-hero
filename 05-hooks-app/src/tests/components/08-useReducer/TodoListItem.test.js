import '@testing-library/jest-dom';
import React from 'react';
import { shallow } from 'enzyme';

import { TodoListItem } from '../../../components/08-useReducer/TodoListItem';
import { demoTodos } from '../../fixtures/demoTodos';

describe('Tests for TodoListItem component', () => {

    const index = 0;
    const demoTodo = demoTodos[index];
    const handleDelete = jest.fn();
    const handleToggle = jest.fn();

    const wrapper = shallow(<TodoListItem
        i={index}
        todo={demoTodo}
        handleDelete={handleDelete}
        handleToggle={handleToggle}
    />);

    test('should match snapshot correctly', () => {        
        expect(wrapper).toMatchSnapshot();
    });

    test('should call delete correclty', () => {
        wrapper.find('button').simulate('click');
        expect(handleDelete).toBeCalledWith(demoTodo.id);
    });

    test('should call toggle correclty', () => {
        wrapper.find('p').simulate('click');
        expect(handleToggle).toBeCalledWith(demoTodo.id);
    });

    test('should print description correctly', () => {
        const text = wrapper.find('p').text().trim();
        expect(text.includes(demoTodo.desc)).toBe(true);
        expect(text.includes((index+1).toString())).toBe(true);
    });


    test('should have complete class if true', () => {
        demoTodo.done = true;

        const wrapper = shallow(<TodoListItem
            i={0}
            todo={demoTodo}
            handleDelete={handleDelete}
            handleToggle={handleToggle}
        />);

        const text = wrapper.find('p');
        expect(text.hasClass('done')).toBe(true);
    });

});