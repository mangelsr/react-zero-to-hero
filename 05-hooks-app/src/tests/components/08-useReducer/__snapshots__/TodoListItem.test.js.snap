// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for TodoListItem component should match snapshot correctly 1`] = `
<li
  className="list-group-item"
>
  <p
    className="false"
    onClick={[Function]}
  >
    1
    . 
    Learn testing
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</li>
`;
